File: Java/Kth-Smallest-Number-in-Sorted-Matrix.java
public class Solution {
    /**
     * @param matrix: a matrix of integers
     * @param k: an integer
     * @return: the kth smallest number in the matrix
     */
    class Point {
        int x, y, val;
        Point() { x = 0; y = 0; val = 0; }
        Point(int a, int b, int c) { x = a; y = b; val = c; }
        Point(Point r) { x = r.x; y = r.y; val = r.val; }
    }
    private Comparator<Point> PointComparator = new Comparator<Point>() {
        public int compare(Point left, Point right) {
            return left.val - right.val;
        }
    };
    public int kthSmallest(int[][] matrix, int k) {
        // write your code here
        int lenx = matrix.length, leny = matrix[0].length;
        PriorityQueue<Point> q = new PriorityQueue<Point>(leny, PointComparator);
        for (int i = 0; i < leny; i ++) {
            q.add(new Point(0, i, matrix[0][i]));
        }
        int res = 0;
        while (!q.isEmpty()) {
            int dx = q.peek().x + 1, dy = q.peek().y, dv = q.peek().val;
            res ++;
            if (res == k) {
                return dv;
            }
            q.poll();
            if (dx < lenx) {
                q.add(new Point(dx, dy, matrix[dx][dy]));
            }
        }
        return -1;
    }
}

File: Cpp/Kth-Smallest-Number-in-Sorted-Matrix.cpp
#include <vector>
#include <queue>

class Solution {
public:
    /**
     * @param matrix: a matrix of integers
     * @param k: an integer
     * @return: the kth smallest number in the matrix
     */
    int kthSmallest(vector<vector<int> > &matrix, int k) {
        // write your code here
        priority_queue<pair<int, pair<int, int> >,vector<pair<int, pair<int, int> > >,
                       greater<pair<int, pair<int, int> > > > que;
        int lenx = matrix.size(), leny = matrix[0].size();
        for (int i = 0; i < leny; i ++) {
            que.push(make_pair(matrix[0][i], make_pair(0, i)));
        }
        int temp, cnt = 0;
        while (cnt < k) {
            pair<int, pair<int, int> > p = que.top();
            que.pop();
            temp = p.first;
            cnt ++;
            int x = p.second.first, y = p.second.second;
            if (x + 1 < lenx) {
                que.push(make_pair(matrix[x+1][y], make_pair(x + 1, y)));
            }
        }
        return temp;
    }
};
